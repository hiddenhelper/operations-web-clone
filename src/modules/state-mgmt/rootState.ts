import { History } from 'history';

import { IState as IAuthState } from './auth/state';
import { IState as IGeneralState } from './general/state';
import { IState as IClientState } from './client/state';
import { IState as IProjectState } from './project/state';
import { IState as IUserState } from './user/state';
import { IState as IAccessControlSystemState } from './access-control-system/state';
import { IState as IBadgePrintingSystemState } from './badge-printer-system/state';
import { IState as IWorkerState } from './worker/state';
import { IState as ICertificationState } from './certification/state';
import { IState as IFileState } from './file/state';
import { IState as IBadgeState } from './badge/state';
import { IState as IStatisticsState } from './statistics/state';
import { IState as ITrainingState } from './training/state';
import { IState as IPaymentState } from './payment/state';
import { IState as IInvoiceState } from './invoice/state';
import { ApiService } from '../services/ApiService';
import { AuthService } from '../services/AuthService';
import { PrinterService } from '../services/PrinterService';
import { FileService } from '../services/FileService';
import { Logger } from '../services/Logger';
import { IProcoreState } from './procore/state';
import { IState as IProjectNewState } from './project-new/state';

export interface IAction {
  type: string;
  payload: any;
}

export interface IRootState {
  auth: IAuthState;
  user: IUserState;
  client: IClientState;
  project: IProjectState;
  accessControlSystem: IAccessControlSystemState;
  badgePrinterSystem: IBadgePrintingSystemState;
  worker: IWorkerState;
  certification: ICertificationState;
  training: ITrainingState;
  invoice: IInvoiceState;
  file: IFileState;
  badge: IBadgeState;
  statistics: IStatisticsState;
  payment: IPaymentState;
  general: IGeneralState;
  router?: { location: { pathname: string; search: string; hard: string; key: string }; action: string };
  procore: IProcoreState;
  projectNew: IProjectNewState;
}

export interface IEpicDependencies {
  history: History;
  apiService: ApiService;
  authService: AuthService;
  printerService: PrinterService;
  fileService: FileService;
  logger: typeof Logger;
}
