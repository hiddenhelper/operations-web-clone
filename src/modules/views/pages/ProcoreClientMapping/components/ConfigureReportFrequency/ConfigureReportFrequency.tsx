import React, { useState, memo, useEffect, useCallback } from 'react';
import { createStyles, makeStyles } from '@material-ui/core/styles';
import { Typography } from '@material-ui/core';
import CreateIcon from '@material-ui/icons/Create';
import { toREM } from 'utils/generalUtils';
import { GeneralModel } from 'modules/models';
import Modal from 'modules/views/shared/Modal';
import Confirm from 'modules/views/shared/Modal/components/Confirm';
import { useStylesModalCards } from 'modules/views/shared/PaymentMethods/styles';
import ControlledInput from 'modules/views/shared/FormHandler/ControlledInput';
import ControlledSelect from 'modules/views/shared/FormHandler/ControlledSelect';

const useStyles = makeStyles(theme =>
  createStyles({
    container: {
      display: 'flex',
      alignItems: 'center',
      cursor: 'pointer',
    },
    text: {
      fontSize: toREM(15),
      fontWeight: 600,
      color: '#444444',
      lineHeight: toREM(24),
      paddingRight: toREM(14),
      textTransform: 'capitalize',
    },
    icon: {
      color: '#878787',
    },
  })
);

const useCustomModalStyles = makeStyles(theme =>
  // TODO: those styles needs to be migrated to a shared compoennt. Shared comoponent needs a refactor soon.
  createStyles({
    title: {
      '& .MuiPaper-root': {
        padding: '20px 20px 16px 20px',
      },
      '& .MuiDialogTitle-root > .MuiTypography-root': {
        height: toREM(32),
        color: 'rgba(0,0,0,0.9)',
        fontFamily: 'Hind',
        fontSize: toREM(20),
        fontWeight: 600,
        letterSpacing: 0,
        lineHeight: toREM(32),
        marginBottom: '8px',
      },
      '& .MuiDialogTitle-root': {
        padding: 0,
      },
      '& .MuiDialogContent-root': {
        padding: 0,
      },
      '& .MuiDialogActions-root .MuiGrid-root .MuiGrid-root:last-child .MuiButtonBase-root': {
        width: '120px',
      },
      '& .MuiDialogActions-root': {
        padding: 0,
      },
    },
    subTitle: {
      height: toREM(46),
      color: '#999999',
      fontFamily: 'Hind',
      fontSize: toREM(15),
      letterSpacing: 0,
      lineHeight: toREM(23),
      marginBotton: toREM(20),
    },
    selectLabel: {
      '& .MuiTypography-root': {
        height: toREM(24),
        color: '#444444',
        fontFamily: 'Hind',
        fontSize: toREM(15),
        letterSpacing: 0,
        lineHeight: toREM(24),
      },
    },
  })
);

export interface IConfigureReportFrequencyProps {
  procoreReportFrequency?: number;
  handleSaveProcoreReportFrequency: (frequency: number) => void;
  procoreSaveReportFrequencyLoading: GeneralModel.ILoadingStatus;
  className?: string;
}

const ConfigureReportFrequency = ({
  procoreReportFrequency,
  handleSaveProcoreReportFrequency,
  procoreSaveReportFrequencyLoading,
  className,
}: IConfigureReportFrequencyProps) => {
  const [showModal, setShowModal] = useState<boolean>(false);
  const [frequencySelected, setFrequencySelected] = useState(1);

  const classes = useStyles();
  const modalClasses = useStylesModalCards();
  const customModalClasses = useCustomModalStyles();

  const frequencies = [
    { value: 0, label: 'Daily' },
    { value: 1, label: 'Daily / Hourly' },
  ];

  const frequencySelectedName = frequencies.find(frequency => frequency.value === procoreReportFrequency);

  useEffect(() => {
    setFrequencySelected(procoreReportFrequency);
  }, [procoreReportFrequency]);

  const handleOpenModal = () => {
    setShowModal(true);
  };

  const handleCloseModal = useCallback(() => setShowModal(() => false), [setShowModal]);

  const handleChangeReportFrequency = /* istanbul ignore next */ e => {
    setFrequencySelected(e.target.value);
  };

  const handleSubmitReportFrequency = () => {
    handleSaveProcoreReportFrequency(frequencySelected);
    handleCloseModal();
  };

  return (
    <>
      <div data-testid="action-configure-report-frequency-container" className={`${classes.container} ${className}`} onClick={handleOpenModal}>
        <span className={classes.text}>
          {procoreSaveReportFrequencyLoading?.isLoading && 'Saving...'}
          {procoreSaveReportFrequencyLoading?.hasError && 'Error Saving'}
          {!procoreSaveReportFrequencyLoading?.isLoading && frequencySelectedName?.label}
        </span>
        <CreateIcon height="18px" className={classes.icon} />
      </div>
      <Modal
        show={showModal}
        styleClass={customModalClasses.title}
        onClose={handleCloseModal}
        render={() => (
          <Confirm
            title="Change Frequency"
            closeLabel="Close"
            confirmLabel="Save"
            onConfirm={handleSubmitReportFrequency}
            onClose={handleCloseModal}
            isLoading={procoreSaveReportFrequencyLoading?.isLoading || false}
            confirmLoadingText="Saving..."
            content={
              <div className={`${modalClasses.replaceModal}`}>
                <Typography className={`${modalClasses.replaceModalText} ${customModalClasses.subTitle}`}>
                  Select the Timeframe in which the Manpower information is sent to Procore.
                </Typography>
                <ControlledInput label="Frequency" styleClass={customModalClasses.selectLabel}>
                  <ControlledSelect
                    name="select-report-frequency"
                    inputProps={{
                      'data-testid': 'select-report-frequency',
                    }}
                    options={frequencies}
                    onChange={handleChangeReportFrequency}
                    disabled={false}
                    value={frequencySelected}
                  />
                </ControlledInput>
              </div>
            }
          />
        )}
      />
    </>
  );
};

export default memo(ConfigureReportFrequency);
