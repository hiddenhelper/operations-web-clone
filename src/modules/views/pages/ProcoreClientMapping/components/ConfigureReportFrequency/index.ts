import ConfigureReportFrequency from './ConfigureReportFrequency';
export { ConfigureReportFrequency };
