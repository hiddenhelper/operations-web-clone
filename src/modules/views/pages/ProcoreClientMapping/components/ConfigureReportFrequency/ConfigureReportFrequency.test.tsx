import React from 'react';
import { Provider } from 'react-redux';
import { createMockStore } from 'redux-test-utils';
import { render, fireEvent, act } from '@testing-library/react';
import { getInitialState } from 'test/rootState';
import ConfigureReportFrequency, { IConfigureReportFrequencyProps } from './ConfigureReportFrequency';
import { getDefaultLoading } from 'test/entities';

describe('Configure Report Frequency', () => {
  describe('With Suggested Mappings', () => {
    let props: IConfigureReportFrequencyProps;
    beforeEach(() => {
      props = {
        handleSaveProcoreReportFrequency: jest.fn(),
        procoreReportFrequency: 0,
        procoreSaveReportFrequencyLoading: getDefaultLoading(),
      };
    });

    it('should render', () => {
      const wrapper = render(
        <Provider store={createMockStore(getInitialState()) as any}>
          <ConfigureReportFrequency {...props} />
        </Provider>
      );

      expect(wrapper.container).toMatchSnapshot();
    });

    it('should open modal and close the modal', () => {
      const wrapper = render(
        <Provider store={createMockStore(getInitialState()) as any}>
          <ConfigureReportFrequency {...props} />
        </Provider>
      );
      act(() => {
        fireEvent.click(wrapper.getByTestId('action-configure-report-frequency-container'));
      });
      expect(wrapper.container).toMatchSnapshot();

      // close the modal
      act(() => {
        fireEvent.click(wrapper.getByTestId('modal-close-btn'));
      });
      expect(wrapper.container).toMatchSnapshot();
    });

    it('should change the report frequency', async () => {
      const wrapper = render(
        <Provider store={createMockStore(getInitialState()) as any}>
          <ConfigureReportFrequency {...props} />
        </Provider>
      );

      act(async () => {
        await fireEvent.click(wrapper.getByTestId('action-configure-report-frequency-container'));
      });

      const frequencySelect: any = await wrapper.getByTestId('select-report-frequency');
      await fireEvent.change(frequencySelect, { target: { value: 1, label: 'Daily / Hourly' } });
      await expect(frequencySelect.label).toBe('Daily / Hourly');
    });

    it('should call save with the right paramters', async () => {
      const wrapper = render(
        <Provider store={createMockStore(getInitialState()) as any}>
          <ConfigureReportFrequency {...props} />
        </Provider>
      );

      act(async () => {
        await fireEvent.click(wrapper.getByTestId('action-configure-report-frequency-container'));
      });
      act(async () => {
        await fireEvent.click(wrapper.getByTestId('modal-confirm-btn'));
      });
      await expect(props.handleSaveProcoreReportFrequency).toHaveBeenCalledWith(0);
    });

    it('should render the word Saving while saving the changes of the modal', () => {
      props.procoreSaveReportFrequencyLoading.isLoading = true;
      const wrapper = render(
        <Provider store={createMockStore(getInitialState()) as any}>
          <ConfigureReportFrequency {...props} />
        </Provider>
      );

      expect(wrapper.findAllByText('Saving...')).toBeDefined();
    });
    it('should render the word Error Saving if errors appears while saving the changes of the modal', () => {
      props.procoreSaveReportFrequencyLoading.hasError = true;
      const wrapper = render(
        <Provider store={createMockStore(getInitialState()) as any}>
          <ConfigureReportFrequency {...props} />
        </Provider>
      );

      expect(wrapper.findAllByText('Error Saving')).toBeDefined();
    });
  });
});
