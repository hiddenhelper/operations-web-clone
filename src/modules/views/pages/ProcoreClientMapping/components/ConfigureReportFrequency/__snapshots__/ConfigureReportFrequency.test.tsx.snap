// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Configure Report Frequency With Suggested Mappings should open modal and close the modal 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="makeStyles-container-43 undefined"
    data-testid="action-configure-report-frequency-container"
  >
    <span
      class="makeStyles-text-44"
    >
      Daily
    </span>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root makeStyles-icon-45"
      focusable="false"
      height="18px"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </div>
</div>
`;

exports[`Configure Report Frequency With Suggested Mappings should open modal and close the modal 2`] = `
<div
  aria-hidden="true"
>
  <div
    class="makeStyles-container-43 undefined"
    data-testid="action-configure-report-frequency-container"
  >
    <span
      class="makeStyles-text-44"
    >
      Daily
    </span>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root makeStyles-icon-45"
      focusable="false"
      height="18px"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </div>
</div>
`;

exports[`Configure Report Frequency With Suggested Mappings should render 1`] = `
<div>
  <div
    class="makeStyles-container-1 undefined"
    data-testid="action-configure-report-frequency-container"
  >
    <span
      class="makeStyles-text-2"
    >
      Daily
    </span>
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root makeStyles-icon-3"
      focusable="false"
      height="18px"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 00-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
      />
    </svg>
  </div>
</div>
`;
